kwasm-lemmas.md

```diff
@@ -20,13 +20,154 @@ Basic logic
+(x mod m) mod n
+ = r mod n
+ = n * ( k * q) + r mod n
+ = m * q + r mod n
+ = x mod n
+```
+
+#### Modulus Over Addition and Division
+
+```k
+    rule (X *Int M +Int Y) modInt N => Y modInt N   requires M modInt N ==Int 0 [simplification]
+    rule (Y +Int X *Int M) modInt N => Y modInt N   requires M modInt N ==Int 0 [simplification]
+```
+
+Proof:
+
+```
+Assume m = k * n for some k (side condition).
+x * m + y mod n = x * (k * n) + y mod n = y mod n
+```
+
+```k
+    rule ((X modInt M) +Int Y) modInt N => (X +Int Y) modInt N   requires M modInt N ==Int 0 [simplification]
+    rule (X +Int (Y modInt M)) modInt N => (X +Int Y) modInt N   requires M modInt N ==Int 0 [simplification]
+```
+
+Proof:
+
+```
+Assume m = l * n
+x = k * m + r, r < m
+x mod m + y = r + y
+(x + y) mod n
+  = k * m + r + y mod n
+  = k * l * n + r + y mod n
+  = r + y mod n
+  = (x mod m + y) mod n
+```
+
+#### Bit Shifting
+
+We want Z3 to understand what a bit-shift is.
+
+```k
+    rule (X >>Int N)          => 0 requires X <Int 2 ^Int N [simplification]
+    rule ( X <<Int N) modInt M => 0 requires M <Int 2 ^Int N [simplification]
+
+    rule (X >>Int N) >>Int M => X >>Int (N +Int M) [simplification]
+    rule (X <<Int N) <<Int M => X <<Int (N +Int M) [simplification]
+
+    // The Haskell and Java backend accept negative shifts (the LLVM and OCaml backends do not).
+    // So removing the side conditions and keeping one of each rule here could give faster symbolic execution.
+    rule (X <<Int N) >>Int M => X <<Int (N -Int M)   requires          N >=Int M  [simplification]
+    rule (X <<Int N) >>Int M => X >>Int (M -Int N)   requires notBool (N >=Int M) [simplification]
+```
+
+The following rules decrease the modulus by rearranging it around a shift.
+
+```k
+    rule (X          modInt POW) >>Int N => (X >>Int N) modInt (POW /Int (2 ^Int N))            requires POW modInt (2 ^Int N) ==Int 0 [simplification]
+    rule (X <<Int N) modInt POW          => (X          modInt (POW /Int (2 ^Int N))) <<Int N   requires POW modInt (2 ^Int N) ==Int 0 [simplification]
+```
+
+Proof sketch: Taking modulo `2^n * k` can only affect the the `n`-th and higher bits.
+A shift right by `n` bits can only erase bits up to the `n-1`-th bit.
+Therefore, we may as well shift first and then take the modulus, only we need to make sure the modulus acts on the shifted bits, by taking modulo `k` instead.
+The argument for the left shift is similar.
+
+```k
+    rule (X +Int (Y <<Int N)) modInt POW => X modInt POW   requires (2 ^Int N) modInt POW ==Int 0 [simplification]
+    rule ((Y <<Int N) +Int X) modInt POW => X modInt POW   requires (2 ^Int N) modInt POW ==Int 0 [simplification]
+```
+
+Proof: These follow from the fact that shifting left by `n` bits is simply multiplying by `2^n`, and from previously proven rules of modular arithmetic.
+
+The next rules consider (some) cases where integers with disjoint 1-bits are added together.
+
+```k
+    rule (X +Int Y) >>Int N => (Y >>Int N)
```

Comment:
is this correct? if X = 3, Y = 7, N = 2,  (X + Y) >> N = 10 / 4 = 2, Y >> N = 7 / 4 = 1.  
In fact, if the side condition is X <Int 2 ^Int (N -Int 1), you can prove the addition causes no carriers.

Reply:
Good catch, you're 100% right.
Turns out this lemma isn't needed for the proof to go through.